presto-tests/src/test/java/io/prestosql/tests/TestQuerySerializationFailures.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.tests;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.prestosql.spi.Plugin;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.AbstractVariableWidthType;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeSignature;
+import io.prestosql.testing.AbstractTestQueryFramework;
+import io.prestosql.testing.DistributedQueryRunner;
+import io.prestosql.testing.QueryRunner;
+import org.testng.annotations.Test;
+
+import java.util.Set;
+
+import static io.prestosql.spi.type.StandardTypes.BOOLEAN;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+
+public class TestQuerySerializationFailures
+        extends AbstractTestQueryFramework
+{
+    @Test
+    public void shouldFailOnFirstSerializationError()
+    {
+        assertQueryFails("SELECT * FROM ( VALUES BOGUS(true), BOGUS(false), BOGUS(true) )", "Could not serialize type 'Bogus' value at position 1:1");
+    }
+
+    @Test
+    public void shouldPass()
+    {
+        assertQuerySucceeds("SELECT * FROM ( VALUES BOGUS(false) )");
+    }
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        DistributedQueryRunner queryRunner = DistributedQueryRunner
+                .builder(testSessionBuilder().build())
+                .setNodeCount(4)
+                .build();
+
+        queryRunner.installPlugin(new BogusPlugin());
+
+        return queryRunner;
+    }
+
+    public static class BogusPlugin
+            implements Plugin
+    {
+        @Override
+        public Iterable<Type> getTypes()
+        {
+            return ImmutableList.of(BogusType.BOGUS_TYPE);
+        }
+
+        @Override
+        public Set<Class<?>> getFunctions()
+        {
+            return ImmutableSet.of(BogusFunctions.class);
+        }
+    }
+
+    public static final class BogusType
```

Comment:
Make `BogusType` a top level test class in `presto-main`, then we can reuse it here (this module depends on the `test-jar` from `presto-main`)